### ISSUE TITLE
Missing resize event when resizing window programmatically on Wayland

### ISSUE THREAD

**Issue by tronical** [rust-windowing/winit#2499]
We've noticed an issue rendering with glutin after resizing a window with wayland. I've reduced it down to a test case via a modified glutin example:

```patch
diff --git a/glutin_examples/examples/window.rs b/glutin_examples/examples/window.rs
index 44c9285..411e639 100644
--- a/glutin_examples/examples/window.rs
+++ b/glutin_examples/examples/window.rs
@@ -7,7 +7,9 @@ use glutin::ContextBuilder;
 
 fn main() {
     let el = EventLoop::new();
-    let wb = WindowBuilder::new().with_title("A fantastic window!");
+    let wb = WindowBuilder::new()
+        .with_title("A fantastic window!")
+        .with_inner_size(glutin::dpi::LogicalSize::new(480, 480));
 
     let windowed_context = ContextBuilder::new().build_windowed(wb, &el).unwrap();
 
@@ -17,9 +19,27 @@ fn main() {
 
     let gl = support::load(&windowed_context.context());
 
+    let mut resize_pending = true;
+    let start_time = std::time::Instant::now();
+
     el.run(move |event, _, control_flow| {
         println!("{:?}", event);
-        *control_flow = ControlFlow::Wait;
+
+        if resize_pending {
+            if start_time.elapsed() > std::time::Duration::from_secs(3) {
+                eprintln!("resizing");
+                windowed_context.window().set_inner_size(glutin::dpi::LogicalSize::new(240, 480));
+                windowed_context.window().request_redraw();
+                resize_pending = false;
+                *control_flow = ControlFlow::Wait;
+            } else {
+                *control_flow = ControlFlow::WaitUntil(
+                    std::time::Instant::now() + std::time::Duration::from_secs(1),
+                );
+            }
+        } else {
+            *control_flow = ControlFlow::Wait;
+        }
 
         match event {
             Event::LoopDestroyed => return,
@@ -29,7 +49,11 @@ fn main() {
                 _ => (),
             },
             Event::RedrawRequested(_) => {
-                gl.draw_frame([1.0, 0.5, 0.7, 1.0]);
+                gl.draw_frame(if resize_pending {
+                    [1.0, 0.5, 0.7, 1.0]
+                } else {
+                    [0.5, 1.0, 1.0, 1.0]
+                });
                 windowed_context.swap_buffers().unwrap();
             }
             _ => (),

```

After about 5 seconds, this examples resizes the window. With x11 that works fine, the window receives a resize event as well as a redraw request.

With wayland running on Linux against Weston, I'm observing different behavior: No resize event is received. A redraw event is received though, and when querying the window for its inner size, it reports the correct new size.

The visual result is that the window decoration is drawn with the new geometry but the rest of the surface with the old geometry, as this screenshot demonstrates:

<img width="257" alt="Screenshot 2022-09-21 at 13 04 24" src="https://user-images.githubusercontent.com/1486/191488798-08bb5e51-ff99-452e-9eec-74e791c1f09a.png">

(note the changed background color, so this isn't an old buffer)

At this point one might think: Ahh, the underlying EGL surface needs to be resized, as if a resize event was received. Manually adding a resize() call on the windowed_context before drawing however does not fix the issue.

I don't know how to work around this and I'm not entirely certain that this is a winit issue - perhaps it's glutin. However given that with x11 the resize event is received and with wayland it isn't, I'm thinking that it's perhaps winit related and am filing it here.

I'd be happy to re-file it elsewhere or try workarounds. Or perhaps there's something wrong with the test case in the way that it is using winit?

**Comment by kchibisov** [comment]
>At this point one might think: Ahh, the underlying EGL surface needs to be resized, as if a resize event was received. Manually adding a resize() call on the windowed_context before drawing however does not fix the issue.

>I don't know how to work around this and I'm not entirely certain that this is a winit issue - perhaps it's glutin. However given that with x11 the resize event is received and with wayland it isn't, I'm thinking that it's perhaps winit related and am filing it here.

>I'd be happy to re-file it elsewhere or try workarounds. Or perhaps there's something wrong with the test case in the way that it is using winit?

On Wayland you resize the EGL surface yourself after getting the size. You can resize only once per swap buffers, if you do that multiple times the latter one won't work until the next buffer swap.

Winit has nothing to do with it. Glutin has also nothing to do with it. It just how Wayland works. See latest glutin for a better example on how to do this if struggle with Wayland.

**Comment by tronical** [comment]
> On Wayland you resize the EGL surface yourself after getting the size.

Note that the example is *not* getting the size, it is not receiving the receive event. `set_inner_size` is called but later no resize event is received for this request. That is precisely the issue AFAICS.

**Comment by kchibisov** [comment]
I think on other platforms `set_inner_size` also not resulting in `Resized` event. I don't see it resulting in resize on other platforms and could be a coincidence here. In general it's true that size changing is async operation, but in wayland world when client controls the window sizes it's sort of sync.

I'll open a separate issue for it then.

**Comment by torokati44** [comment]
> See latest glutin for a better example on how to do this if struggle with Wayland.

@kchibisov Could you please help me out with a more precise pointer to this?

**Comment by kchibisov** [comment]
See https://github.com/rust-windowing/winit/issues/2529 . Though, your issue was handled separately, given that you've opened another issue.

I still don't think that always sending `Resized` is a good idea, because it doesn't make any sense at all for sync Wayland case. Maybe we should send a **synthetic** resized event, so you'll have resize, but `synthetic` part will mean that it was applied already for you and you can't do anything about it.
